ocaml/typing/typedecl.ml

```diff
@@ -1113,40 +1171,119 @@ let update_decl_jkind env dpath decl =
-      let has_values, has_floats =
-        Array.fold_left
-          (fun (values, floats) jkind ->
-             match Jkind.get_default_value jkind with
-             | Value | Immediate64 | Immediate -> (Has_values, floats)
-             | Float64 -> (values, Has_float64s)
-             | Void -> (values, floats)
-             (* CR layouts v2.1: make unboxed ints work with records *)
-             | Word | Bits32 | Bits64 ->
-              Misc.fatal_error "Typedecl.update_record_kind: no support for unboxed ints"
-             | Any -> assert false)
-          (No_values, No_float64s) jkinds
+      let classify (lbl : Types.label_declaration) jkind =
+        if is_float env lbl.ld_type then Float_element
+        else match Jkind.get_default_value jkind with
+          | Value | Immediate64 -> Value_element
+          | Immediate -> Flat_imm_element
+          | Float64 -> Flat_float64_element
+          | Void -> Element_without_runtime_component
+          | Word | Bits32 | Bits64 ->
+              Misc.fatal_error
+                "Typedecl.update_record_kind: no support for unboxed ints"
+          | Any ->
+              Misc.fatal_error "Typedecl.update_record_kind: unexpected Any"
+      in
+      let classifications =
+        List.mapi (fun i lbl -> lbl, classify lbl jkinds.(i)) lbls
       in
+      let element_reprs =
+        { values = false; imms = false; floats = false; float64s = false }
+      in
+      List.iter
+        (fun (_lbl, repr) ->
+           match repr with
+           | Float_element -> element_reprs.floats <- true
+           | Flat_imm_element -> element_reprs.imms <- true
+           | Flat_float64_element -> element_reprs.float64s <- true
+           | Value_element -> element_reprs.values <- true
+           | Element_without_runtime_component -> ())
+        classifications;
       let rep =
-        match has_values, has_floats with
-        | Has_values, Has_float64s -> raise (Error (loc, Mixed_block))
-        | Has_values, No_float64s -> rep
-        | No_values, Has_float64s -> Record_ufloat
-        | No_values, No_float64s ->
+        match[@warning "+9"] element_reprs with
+        | { values = false; imms = false; floats = true ; float64s = true  } ->
+            (* We store mixed float/float64 records as flat if there are no
+                non-float fields.
+            *)
+            let flat_suffix =
+              List.map
+                (fun (_, c) ->
+                  match c with
+                  | Float_element -> Float
+                  | Flat_float64_element -> Float64
+                  | Element_without_runtime_component -> Imm
+                  | Flat_imm_element | Value_element ->
+                      Misc.fatal_error "Expected only floats and float64s")
+                classifications
+              |> Array.of_list
+            in
+            assert_mixed_record_support loc ~value_prefix_len:0;
+            Record_mixed { value_prefix_len = 0; flat_suffix }
```

Comment:
Why do we use a mixed record there instead of a flat float record with `tag = 252`?

Reply:
First, why even support this case at all? The thinking is that this gives users a way of gradually migrating their all-float records to all-unboxed-floats without sacrificing the flat float optimization at any point.

Next, what tag to use?
  * Downside of current approach: the user loses polymorphic operations. But this is fine for Jane Street code.
  * Benefit of current approach: this could allow us to store boxed floats as flat even in inline record arguments to constructors. But actually I don't plan on supporting this: The argument for flat floats in mixed records doesn’t apply, as no existing code can possibly rely on the flat float optimization in constructor arguments.

So it does seem a bit better to use `Double_array_tag` here. The main reason I didn't do that in this PR is that @ccasin tells me it was hard to teach the middle-end how to do projections correctly here (some require boxing and others don't), and would require a small refactor.

But I plan on looking into this (in another PR) and suggest going with this approach for now.